libs/ebpfnetsh/processes.cpp

```diff
@@ -0,0 +1,223 @@
+#include <processsnapshot.h>
+#include <psapi.h>
+#pragma comment(lib, "psapi.lib")
+#include <stdio.h>
+#include <winternl.h>
+#pragma comment(lib, "ntdll.lib")
+
+#define ObjectNameInformation ((OBJECT_INFORMATION_CLASS)1)
+typedef struct
+{
+    UNICODE_STRING Name;
+    WCHAR NameBuffer[1];
+} OBJECT_NAME_INFORMATION;
+
+typedef struct
+{
+    HANDLE target_handle;
+    WCHAR name[MAX_PATH];
+} query_name_param_t;
+
+// Get the object name associated with a target handle.
+static DWORD WINAPI
+_query_name_thread_proc(_In_ void* parameter)
+{
+    query_name_param_t* param = (query_name_param_t*)parameter;
+
+    char buffer[1024];
+    ULONG buffer_size = sizeof(buffer);
+    NTSTATUS status;
+    OBJECT_NAME_INFORMATION* object_name_info = (OBJECT_NAME_INFORMATION*)buffer;
+    status = NtQueryObject(param->target_handle, ObjectNameInformation, object_name_info, buffer_size, nullptr);
+    if (NT_SUCCESS(status) && (object_name_info->Name.Length > 0)) {
+        memcpy(param->name, object_name_info->Name.Buffer, object_name_info->Name.Length);
+        param->name[object_name_info->Name.Length / sizeof(WCHAR)] = 0;
+    } else {
+        param->name[0] = 0;
+    }
+    return 0;
+}
+
+static void
+_print_process_info(HANDLE process_handle)
+{
+    WCHAR process_name[MAX_PATH];
+    (void)GetModuleBaseName(process_handle, nullptr, process_name, sizeof(process_name) / sizeof(*process_name));
+    std::wcout << std::right << std::setw(5) << GetProcessId(process_handle) << "  " << std::wstring(process_name)
+               << std::endl;
+}
+
+// Print info on handle if we can get the name of the object it references.
+static void
+_print_handle_info(HANDLE process_handle, HANDLE other_handle)
+{
+    // Duplicate the handle into the local process so we can query it.
+    HANDLE local_handle;
+    if (!DuplicateHandle(
+            process_handle, other_handle, GetCurrentProcess(), &local_handle, 0, FALSE, DUPLICATE_SAME_ACCESS)) {
+        // Some handle types cannot be duplicated. This is ok, since the handles we're looking for can be
+        // duplicated.  Also, we can't duplicate handles in some system security processes. This is ok,
+        // we just can't detect whether they have handles open of the type we're looking for.
+        return;
+    }
+
+    // Get type information.
+    query_name_param_t param = {};
+    char buffer[1024];
+    ULONG buffer_size = sizeof(buffer);
+    PUBLIC_OBJECT_TYPE_INFORMATION object_type_info;
+    ULONG bytes_needed = 0;
+    HANDLE thread;
+    NTSTATUS status = NtQueryObject(local_handle, ObjectTypeInformation, buffer, buffer_size, &bytes_needed);
+    if (!NT_SUCCESS(status)) {
+        goto Done;
+    }
+    memcpy(&object_type_info, buffer, sizeof(object_type_info));
+
+    // eBPF handles are always of type "File".
+    if (wcscmp(object_type_info.TypeName.Buffer, L"File") != 0) {
+        goto Done;
+    }
+
+    // Set a 200ms timeout for querying object name after which we just give up.
```

Comment:
File an issue to track this improvement.

Reply:
Where?  (btw, resource manager does the same thing here)

Reply:
I was proposing filing an GH issue to track using a thread pool work item instead of spawning a thread each time.